src/azure-cli-core/azure/cli/core/auth/binary_cache.py

```diff
@@ -0,0 +1,84 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import collections.abc as collections
+import pickle
+
+from azure.cli.core.decorators import retry
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class BinaryCache(collections.MutableMapping):
+    """
+    Derived from azure.cli.core._session.Session.
+    A simple dict-like class that is backed by a binary file.
+
+    All direct modifications with `__setitem__` and `__delitem__` will save the file.
+    Indirect modifications should be followed by a call to `save`.
+    """
+
+    def __init__(self, file_name):
+        super().__init__()
+        self.filename = file_name
+        self.data = {}
+        self.load()
+
+    @retry()
+    def _load(self):
+        """Load cache with retry. If it still fails at last, raise the original exception as-is."""
```

Comment:
Nit-picking: This method does not currently retry. But that is perhaps good enough. We would just need to rephrase this comment.